sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/group/ThroughputGroupControllerBase.java

```diff
@@ -0,0 +1,238 @@
+        checkNotNull(cancellationToken, "Cancellation token can not be null");
+        return Mono.delay(DEFAULT_THROUGHPUT_USAGE_RESET_DURATION)
+            .flatMap(t -> this.resolveRequestController())
+            .flatMap(requestController -> requestController.renewThroughputUsageCycle(this.groupThroughput.get()))
+            .onErrorResume(throwable -> {
+                logger.warn("Reset throughput usage failed with reason %s", throwable);
+                return Mono.empty();
+            }).repeat(() -> !cancellationToken.isCancellationRequested());
+    }
+
+    private Mono<IThroughputRequestController> createAndInitializeRequestController() {
+        IThroughputRequestController requestController;
+        if (this.connectionMode == ConnectionMode.DIRECT) {
+            requestController = new PkRangesThroughputRequestController(
+                this.globalEndpointManager,
+                this.partitionKeyRangeCache,
+                this.targetContainerRid,
+                this.groupThroughput.get());
+
+        } else if (this.connectionMode == ConnectionMode.GATEWAY) {
+            requestController = new GlobalThroughputRequestController(this.globalEndpointManager, this.groupThroughput.get());
+        } else {
+            throw new IllegalArgumentException(String.format("Connection mode %s is not supported"));
+        }
+
+        return requestController.init();
+    }
+
+    public boolean isUseByDefault() {
+        return this.group.isUseByDefault();
+    }
+
+    public void onContainerMaxThroughputRefresh(int maxContainerThroughput) {
+        if (this.maxContainerThroughput.getAndSet(maxContainerThroughput) != maxContainerThroughput) {
+            this.calculateGroupThroughput();
+        }
+    }
+
+    @Override
+    public Mono<Void> close() {
+        this.cancellationTokenSource.cancel();
+        return this.resolveRequestController()
+            .flatMap(requestController -> requestController.close());
+    }
+
+    @Override
+    public <T> Mono<T> processRequest(RxDocumentServiceRequest request, Mono<T> nextRequestMono) {
+        return this.resolveRequestController()
+            .flatMap(requestController -> {
+                if (requestController.canHandleRequest(request)) {
+                    return Mono.just(requestController);
+                } else {
+                    // We can not find the matching pkRange, it is because either the group control is out of sync
+                    // or the request has staled info.
+                    // We will handle the first scenario by creating a new request controller,
+                    // while for second scenario, we will go to the original request mono, which will eventually get exception from server
+                    return this.shouldUpdateRequestController(request)
+                        .flatMap(shouldUpdate -> {
+                            if (shouldUpdate) {
+                                requestController.close();
```

Comment:
Again, no one is subscribing this `close()` call, so this won't get executed at all, hence leaking resources.

Reply:
gotcha, will update in the next iterations.

Reply:
Updated